src/analysis/confusion_matrix.rs

```diff
@@ -0,0 +1,182 @@
+/// predicted to lie in class j but actually lie in class i. (i.e.
+/// the row is the true class and the column is the predicted.).
+///
+/// # Arguments
+/// * `predictions` - A series of model predictions.
+/// * `targets` - A slice of equal length to predictions, containing the target results.
+/// * `labels` - If not provided then the returned matrix rows and columns correspond to
+///              the distinct elements of union(predictions, targets), in increasing
+///              order. If labels are provided then these are used instead, in the order
+///              provided. Note that this means that the confusion matrix will only contain
+///              entries for the elements of labels.
+///
+/// # Examples
+/// ```
+/// use rusty_machine::analysis::confusion_matrix::confusion_matrix;
+/// use rusty_machine::linalg::Matrix;
+///
+/// let truth       = vec![2, 0, 2, 2, 0, 1];
+/// let predictions = vec![0, 0, 2, 2, 0, 2];
+///
+/// let confusion = confusion_matrix(&predictions, &truth, None);
+///
+/// let expected = Matrix::new(3, 3, vec![
+///     2, 0, 0,
+///     0, 0, 1,
+///     1, 0, 2]);
+///
+/// assert_eq!(confusion, expected);
+/// ```
+/// # Panics
+///
+/// - If user-provided labels are not distinct.
+/// - If predictions and targets have different lengths.
+pub fn confusion_matrix<T>(predictions: &[T],
+                           targets: &[T],
+                           labels: Option<Vec<T>>) -> Matrix<usize>
+    where T: Ord + Eq + Hash + Copy
+{
+    assert!(predictions.len() == targets.len(),
+        "predictions and targets have different lengths");
+
+    let labels = match labels {
+        Some(ls) => ls,
+        None => ordered_distinct(predictions, targets)
+    };
+
+    let mut label_to_index: HashMap<T, usize> = HashMap::new();
+    for (i, l) in labels.iter().enumerate() {
+        match label_to_index.insert(*l, i) {
+            None => {},
+            Some(_) => { panic!("labels must be distinct"); }
+        }
+    }
+
+    let mut counts = Matrix::new(labels.len(), labels.len(),
+        vec![0usize; labels.len() * labels.len()]);
+
+    for (truth, pred) in targets.iter().zip(predictions) {
+        if label_to_index.contains_key(truth) && label_to_index.contains_key(pred) {
+            let row = label_to_index[truth];
+            let col = label_to_index[pred];
+
+            counts[[row, col]] += 1;
+        }
+    }
+
+    counts
+}
+
+fn ordered_distinct<T: Ord + Eq + Copy>(xs: &[T], ys: &[T]) -> Vec<T> {
+    let mut ds: Vec<T> = xs.iter().chain(ys).map(|x| *x).collect();
+    ds.sort();
+    ds.dedup();
+    ds
+}
+
+#[cfg(test)]
+mod tests {
+    use super::confusion_matrix;
+    use linalg::Matrix;
+
+    #[test]
+    fn confusion_matrix_no_labels() {
+        let truth       = vec![2, 0, 2, 2, 0, 1];
+        let predictions = vec![0, 0, 2, 2, 0, 2];
+
+        let confusion = confusion_matrix(&predictions, &truth, None);
+
+        let expected = Matrix::new(3, 3, vec![
```

Comment:
In rulinalg we're trying to follow the convention of using the new [`matrix!`](https://athemathmo.github.io/rulinalg/doc/rulinalg/macro.matrix.html) macro for any new tests. It's a minor point but I thought I'd bring it up as it is a _little_ bit more readable than the `Matrix::new` function with a `vec!` macro.